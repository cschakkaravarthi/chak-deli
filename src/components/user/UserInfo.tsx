import React, { FC } from 'react';
import get from 'lodash.get';
import UserInfoModel, { Phone } from '../../shared/models/UserInfo.model';

type Props = {
  usersInfo?: UserInfoModel;
};

const UserInfo: FC<Props> = props => {
  const { usersInfo } = props;

  if (!usersInfo || !usersInfo.networkID) return null;

  return (
    <>
      <div className='user__info'>
        <div className='user__name'>
          <a href='#'>
            {usersInfo.lastName}, {usersInfo.firstName}
          </a>
        </div>
        <div className='user__title'>
          <p>{usersInfo.jobTitle}</p>
          <p>{usersInfo.department}</p>
        </div>
        <div className='user__email'>
          <a type='email' href={`mailto:${usersInfo.email}`}>
            {usersInfo.email}
          </a>
        </div>
        {get(usersInfo, 'location.office') && (
          <div className='user__office'>
            <p>Office: {get(usersInfo, 'location.office', '')}</p>
          </div>
        )}
        <div className='user__address'>
          <p>
            {get(usersInfo, 'location.room', '')} <br />
            {get(usersInfo, 'location.address', '')} <br />
            {`${get(usersInfo, 'location.city', '')}, ${get(usersInfo, 'location.state', '')} `}
            {get(usersInfo, 'location.postalCode', '')}
          </p>
        </div>
        <div className='user__phone'>
          {get(usersInfo, 'phone', [] as Phone[]).map((item: Phone, index: number) => (
            <p key={index}>
              {item.type === 'Desk' && <span>Phone: {item.num}</span>}
              {item.type === 'Mobile' && <span>Mobile: {item.num}</span>}
            </p>
          ))}
        </div>
      </div>
    </>
  );
};

export default UserInfo;
