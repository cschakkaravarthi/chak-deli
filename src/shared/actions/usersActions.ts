import { ThunkAction, ThunkDispatch } from 'redux-thunk';

import { AnyAction } from 'redux';

import {
  LoadUsersAction,
  LoadUserDetailsAction,
  LOAD_USERS,
  FETCH_USERS,
  SET_USER_DETAILS
} from '../types/userTypes';

import ApiService from '../services/apiService';
import UserInfoModel from '../models/UserInfo.model';

export const loadUsers = (users: UserInfoModel[]): LoadUsersAction => ({
  users,
  type: LOAD_USERS
});

export const setUserDetails = (user: UserInfoModel): LoadUserDetailsAction => ({
  user,
  type: SET_USER_DETAILS
});

export const fetchUserDetails = (
  userID: string
): ThunkAction<Promise<void>, {}, {}, AnyAction> => {
  return async (dispatch: ThunkDispatch<{}, {}, AnyAction>): Promise<void> => {
    dispatch(() => ({ type: FETCH_USERS }));
    return ApiService.getUserByNetworkId(
      user => dispatch(setUserDetails(user[0])),
      userID
    );
  };
};

export const fetchUsersByEmail = (
  email: string
): ThunkAction<Promise<void>, {}, {}, AnyAction> => {
  return async (dispatch: ThunkDispatch<{}, {}, AnyAction>): Promise<void> => {
    dispatch(() => ({ type: FETCH_USERS }));
    return ApiService.getUserByEmail(
      user => dispatch(loadUsers(user)),
      email
    );
  };
};
