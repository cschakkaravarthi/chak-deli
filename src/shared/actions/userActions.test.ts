import moxios from 'moxios';
import { dummyUserInfo } from '../models/UserInfo.model';
import { loadUsers, fetchUserDetails } from './usersActions';
import getStore from '../services/mockGlobalStore';

describe('userActions', () => {
  let store: any;
  beforeEach(() => {
    store = getStore();
    moxios.install();
  });

  afterEach(() => {
    moxios.uninstall();
  });

  it('creates a loadUsers action', () => {
    expect(loadUsers([dummyUserInfo])).toMatchSnapshot();
  });

  it.skip('fetchUserDetails makes API call', done => {
    const expectedActions = ['SET_USER_DETAILS'];

    moxios.stubRequest('user', {
      status: 200,
      responseText: 'OK'
    });

    store
      .dispatch(fetchUserDetails('AbedA'))
      .then(() => {
        const actionsCalled = store
          .getActions()
          .map((action: any) => action.type);
        expect(actionsCalled).toEqual(expectedActions);
        done();
      })
      .catch((error: any) => {
        done(error);
      });
  });
});
