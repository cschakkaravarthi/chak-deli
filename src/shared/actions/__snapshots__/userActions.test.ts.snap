// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`userActions creates a loadUsers action 1`] = `
Object {
  "type": "LOAD_USERS",
  "users": Array [
    Object {
      "department": "UMG Corporate IT",
      "directReports": Array [
        Object {
          "firstName": "Artrina",
          "lastName": "Hines",
          "networkID": "HinesAr",
        },
        Object {
          "firstName": "Ann",
          "lastName": "Vanderberg",
          "networkID": "VanderA",
        },
        Object {
          "firstName": "Apollo",
          "lastName": "Capito",
          "networkID": "CapitoA",
        },
        Object {
          "firstName": "Christopher",
          "lastName": "Gillis",
          "networkID": "GillisC",
        },
      ],
      "email": "doris.aves@umusic.com",
      "error": Object {
        "errorDetails": Object {},
        "isError": false,
        "message": "",
      },
      "firstName": "Luz",
      "jobTitle": "Senior Director",
      "lastName": "Lacson",
      "location": Object {
        "address": "21301 Burbank Blvd",
        "city": "Woodland Hills",
        "country": "US",
        "office": "Woodland Hills - UMG",
        "postalCode": "91367",
        "room": "A-31",
        "state": "CA",
      },
      "networkID": "AvesD",
      "phone": Array [
        Object {
          "cc": "US",
          "num": "+1 221 397 5664",
          "type": "Desk",
        },
        Object {
          "cc": "US",
          "num": "+1 413 286 671",
          "type": "Mobile",
        },
      ],
      "photo": Object {
        "large": "test-file-stub",
        "medium": "test-file-stub",
        "small": "test-file-stub",
        "thumb": "test-file-stub",
      },
      "preferredLanguage": "en-US",
      "profilePicture": "",
      "reportsTo": Object {
        "department": "UMG Corp IT/R&C Run",
        "firstName": "Andre",
        "jobTitle": "Vice President, IT Run/Operations",
        "lastName": "Abed",
        "networkID": "AbedA",
      },
    },
  ],
}
`;
